design/design-doc-composition.md

```diff
@@ -0,0 +1,816 @@
+namespaced infrastructure resource (`KubernetesCluster`) that allows application
+operators to bind two cluster scoped infrastructur resources (a `GKECluster` and
+a `NodePool`).
+
+### Transform Functions
+
+TODO(negz): Explain transform functions.
+
+* Simple, schemafied, functions that can transform claim fields.
+* Similar to Terraform's small set of functions.
+* Purposefully not open ended.
+* Written using plain old YAML.
+
+### Composite References
+
+Crossplane allows certain fields of primitive managed resources to be set to a
+value inferred from another primitive managed resource using [cross resource
+references]. Take for example a resource that must specify the VPC network in
+which it should be created, by specifying the `network` field:
+
+```yaml
+spec:
+  forProvider:
+    network: /projects/example/global/networks/desired-vpc-network
+```
+
+Cross resource references allow this resource to instead reference the `Network`
+managed resource that represents the desired VPC Network:
+
+```yaml
+spec:
+  forProvider:
+    networkRef:
+      name: desired-vpc-network
+```
+
+The managed resource reconcile logic resolves this reference and populates the
+`network` field, resulting in the following configuration:
+
+```yaml
+spec:
+  forProvider:
+    # Network is populated with the value calculated by networkRef.
+    network: /projects/example/global/networks/desired-vpc-network
+    networkRef:
+      name: desired-vpc-network
+```
+
+This functionality must be extended in order to support composite resource
+classes. Consider a `Composite` that may be used to dynamically provision the
+following primitive managed resources:
+
+* `Subnetwork` A
+* `GKECluster` A
+* `ServiceAccount` A
+* `ServiceAccount` B
+* `GKENodePool` A
+* `GKENodePool` B
+
+The `Composite` author would like to configure the resources such that:
+
+1. `Subnetwork` A is created in an existing, statically provisioned `Network`.
+1. `GKECluster` A is created in `Subnetwork` A.
+1. `GKENodePool` A uses `ServiceAccount` A.
+1. `GKENodePool` B uses `ServiceAccount` B.
+1. Both `GKENodePool` resources join `GKECluster` A.
+
+The author cannot use a contemporary cross resource reference for requirements
+two through five. Managed resources are referenced by name, and the names of
+dynamically provisioned resources are non-deterministic; they are not known
+until they have been provisioned.
+
+This document proposes the introduction of a reference _selector_, which allows
+a managed resource to describe the properties of the distinct resource it wishes
+to reference, rather than explicitly naming it.
+
+```yaml
+spec:
+  forProvider:
+    networkSelector:
+      # Match only managed resources that are part of the same composite, i.e.
+      # managed resources that have the same controller reference as the
+      # selecting resource.
+      matchController: true
+      # Match only managed resources with the supplied labels.
+      matchLabels:
+        example: label
```

Comment:
>  I think enumerating each base with an identifier and using them in selectors is simpler UX. Not huge difference though.

I prefer this more explicit approach here, even if it is slightly more complicated. The approach in which there is an identifier that gets translated to a label selector "behind the scenes" seems more opaque and thus [more astonishing](https://en.wikipedia.org/wiki/Principle_of_least_astonishment). This approach has you model and consume labels in a common Kubernetes pattern, which is only slightly more verbose. It also means whatever controller handles this templating can be completely unaware of references - it just creates these resources and they automatically relate to each other (because the functionality is built into their controllers).